gaia-header.js

```diff
@@ -145,13 +327,172 @@ module.exports = Component.register('gaia-header', {
    * @param  {Event} e
    * @private
    */
-  onActionButtonClick: function(e) {
-    var config = { detail: { type: this.getAttribute('action') } };
-    var actionEvent = new CustomEvent('action', config);
-    setTimeout(this.dispatchEvent.bind(this, actionEvent));
+  onActionButtonClick: function() {
+    debug('action button click');
+    var config = { detail: { type: this.action } };
+    var e = new CustomEvent('action', config);
+    setTimeout(() => this.dispatchEvent(e));
+  },
+
+  /**
+   * Get the margin required to center
+   * the title between surrounding buttons.
+   *
+   * We don't do any DOM writes in this function
+   * so that we can batch them together with
+   * other write for optimum performance.
+   *
+   * @param  {Object} title
+   * @return {Object}
+   * @private
+   */
+  getTitleMarginStart: function() {
+    var titleStart = this.titleStart;
+    var titleEnd = this.titleEnd;
+    debug('get title margin', titleStart, titleEnd);
+    return titleStart === titleEnd ? 0 : titleEnd - titleStart;
+  },
+
+  /**
+   * Get all the buttons (<a> & <button>)
+   * before the first <h1>.
+   *
+   * @return {Array}
+   * @private
+   */
+  getButtonsBeforeTitle: function() {
+    var els = [];
+
+    for (var i = 0; i < this.children.length; i++) {
+      var el = this.children[i];
+      if (el.tagName === 'H1') { break; }
+      if (!isButton(el.tagName)) { continue; }
+      els.push(el);
+    }
+
+    // Don't forget the action button
+    if (this.action) { els.push(this.els.actionButton); }
+    return els;
+  },
+
+  /**
+   * Get all the buttons (<a> & <button>)
+   * after the last <h1>.
+   *
+   * @return {Array}
+   * @private
+   */
+  getButtonsAfterTitle: function() {
+    var els = [];
+
+    for (var i = this.children.length - 1; i >= 0; i--) {
+      var el = this.children[i];
+      if (el.tagName === 'H1') { break; }
+      if (!isButton(el.tagName)) { continue; }
+      els.push(el);
+    }
+
+    return els;
+  },
```

Comment:
I'm concerned by these 2 functions; are you sure buttons visually before the title are also before in the markup? same question for buttons "after" of course :)

From what I understand about how gaia-header works and how it's used, actually the left buttons are only the "action" button, while the right buttons are the buttons contributed by the header user. Maybe it would make more sense to use this logic instead. What do you think?